k2/csrc/properties.cc

```diff
@@ -78,6 +90,78 @@ bool HasSelfLoops(const Fsa &fsa) {
   return false;
 }
 
+bool CheckCycles(int32_t s, std::vector<bool> *visited,
+                 std::vector<bool> *back_arc, const std::vector<Arc> &arcs) {
+  if ((*visited)[s] == false) {
+    (*visited)[s] = true;
+    (*back_arc)[s] = true;
+
+    std::unordered_set<int32_t> adj;
+    for (const auto &arc : arcs)
+      if (arc.src_state == s)
+        adj.insert(arc.dest_state);
+
+    for (auto i = adj.begin(); i != adj.end(); ++i) {
+      if (!(*visited)[*i] && CheckCycles(*i, visited, back_arc, arcs))
+        return true;
+      else if ((*back_arc)[*i])
+        return true;
+    }
+  }
+  (*back_arc)[s] = false;
+  return false;
+}
+
+// Detect cycles using DFS traversal
+bool IsAcyclic(const Fsa &fsa, std::vector<int32_t> *order /*= nullptr*/) {
```

Comment:
BTW, I just found any approach of `DFS` starting with the start state holds an assumption that the fsa is connected. It's OK in `ConnectCore` as we would only keep connected states in `ConnectCore`. However, this is not true in other situations such as `IsAcyclic`. Maybe implementing this function with `Topological Sorting`(a queue to maintain states with `out-degree == 0`) is straightforward and simpler than `scc`.

Guys, what do you think? I was thinking we may need to avoid to write so much `DFS` or `SCC` code (as they are nearly same structure), it will hard to maintain such code.  Maybe we should establish `openfst`-like structure for such structure(and only for such structrue), i.e. define a DFS structrue and provide different `next/done/visiting/visited` function for different operations.

And, seems sometime it will be very simple if we could maintain a `GetEnterArcs`.

Reply:
I don't think we care about cycles in parts of the FSA that are not connected.  We could just document that they won't be found.Also an "IsAcyclicAndConnected()" function might be useful because in practice people will want both of those conditions at the same time.

Reply:
Regarding a special structure for DFS: I doubt it will be needed except in these two specific instances.  I'm fine to leave that for later or never; I don't think it's a big priority.